### PR TITLE
browser: support client certificate

### LINKED ISSUE #1221: Client certificate not working
Hi,

The docs list `--ignore-certificate-errors` as a supported chromium switch. I'm trying to set up a test/integration/development environment for my apps using self-signed certificates, but am not having any luck getting this switch to work. I'm using v0.21.3 on both Linux and OS X.

My code opens BrowserWindow content fine when pointed at servers with valid certs. But for GET requests to servers without valid certs, the DevTools always show that the network request never receives any bytes. I'm assuming this (silent) failure is a result of ssl validation failure inside chromium.

I've tried both:

```
var app = require ('app');
app.commandLine.appendSwitch ('ignore-certificate-errors');
```

```
var app = require ('app');
app.commandLine.appendSwitch ('ignore-certificate-errors', 'true');
```

Searching the github repo for the strings 'ignore-certificate-errors' and kIgnoreCertificateErrors doesn't produce any code results. I probably just don't understand how command line switches are wired up (I'll look more deeply), but I was wondering if this switch is, in fact, supported and whether the docs are up to date.

Thanks very much.

### PR DESCRIPTION
**Pull request by deepak1556**
Fixes #1221 

- [ ] Tests

### CODE REVIEW

**Comment by zcbenz** [comment]
It seems that Chrome reads from [cert_request_info->client_certs](https://code.google.com/p/chromium/codesearch#chromium/src/chrome/browser/chrome_content_browser_client.cc&sq=package:chromium&type=cs&l=1800&rcl=1433748798) instead of letting user choose a custom file?

**Comment by deepak1556** [comment]
> It seems that Chrome reads from cert_request_info->client_certs instead of letting user choose a custom file?

yup that is the default behaviour where a user selects from the dialog populated with certificates from the system store, but i am trying to address 1 & 2 [here](https://github.com/atom/electron/issues/1221#issuecomment-107699881) . So if a user doesnt provide local client certificate , will fallback to the default behaviour (which havnt handled yet, because of unimplemented table dialog view).

**Comment by zcbenz** [comment]
We can choose the first matched certificate as default behavior, so users don't have to do anything to get client certificate support for free.

For further support when a certificate is requested we can probably emit a new event with available certificates and a callback to pass the selected one, this can be done in another PR.

**Comment by deepak1556** [comment]
If local certificate path is not provided, will default to first certificate from the cert store currently.

Depends on https://github.com/atom/brightray/pull/115

**Comment by deepak1556** [comment]
@zcbenz for default auto-selection , user needs to provide a certificate filter . Currently i am thinking of following chrome's web preference filter [pattern](https://code.google.com/p/chromium/codesearch#chromium/src/components/content_settings/core/browser/content_settings_policy_provider.cc&sq=package:chromium&rcl=1433763137&l=284) . Thoughts ?

```
  {
     "pattern": "[*.]example.com",
     "filter": {
        "ISSUER": {
          "CN": "some name"
        }
     }
   }
```

**Comment by kjlaw89** [comment]
That looks good to me, definitely simple enough.

**Comment by thojansen** [comment]
:+1: looks good. ISSUER.CN would be the most important value for us

File: atom/browser/atom_browser_client.cc [hunk 1]
```diff
@@ -189,6 +209,28 @@ content::QuotaPermissionContext*
...
+  // in the list.
+  scoped_refptr<net::X509Certificate> certificate;
+  if (cert_path.empty()) {
+    if (!cert_request_info->client_certs.empty())
+      certificate = cert_request_info->client_certs[0];
+  } else {
+    certificate = ImportCertFromFile(base::FilePath(cert_path));
+  }
+
+  delegate->ContinueWithCertificate(certificate.get());
```
Comment (zcbenz) [inline atom/browser/atom_browser_client.cc]: Should check for the case when `certificate` is empty.

**Comment by zcbenz** [comment]
> @zcbenz for default auto-selection , user needs to provide a certificate filter . Currently i am thinking of following chrome's web preference filter pattern . Thoughts ?

We can provide an API to make it possible to choose which certificate to use in JavaScript, so developers can either filter out the one they want, or show the available certificates in GUI to let final users decide.

```coffeescript
app.setClientCertificateSelector (webContents, sslCertRequestInfo, callback) ->
  callback(sslCertRequestInfo.certificates[0]);

File: atom/browser/atom_browser_client.cc (hunk 1 above)
Comment (deepak1556) [inline atom/browser/atom_browser_client.cc]: should it be checked ? if the certificate is empty the auth will [abort](https://code.google.com/p/chromium/codesearch#chromium/src/content/browser/ssl/ssl_client_auth_handler.cc&sq=package:chromium&rcl=1433820770&l=149).

Comment (zcbenz) [inline atom/browser/atom_browser_client.cc]: Calling `ContinueWithCertificate` with `nullptr` will finally reach here:
https://code.google.com/p/chromium/codesearch#chromium/src/net/url_request/url_request_job.cc&sq=package:chromium&type=cs&l=226

The code you were referencing checks for `handle` not `cert`.

Comment (deepak1556) [inline atom/browser/atom_browser_client.cc]: fixed.

**Comment by zcbenz** [comment]
:+1:

**Comment by kjlaw89** [comment]
Hey @deepak1556, now that this was released I was able to give it a shot today. It definitely works selecting the first certificate from the store, but I was unable to get the command line switch to work how I would expect. Here's the code I was using:

    var app = require ('app');
    app.commandLine.appendSwitch ('client-certificate', process.cwd () + '\\mycertificate.p12');
    console.log (process.cwd () + "\\mycertificate.p12");

I did verify that the path to the certificate is correct. The p12 actually contains a password, which I'm guessing this method doesn't support, but I did do a .pem that does not have a password and I know works but still did not have any luck. Am I missing something?

**Comment by deepak1556** [comment]
@kjlaw89 yup this method currently supports only PEM encoded data, but your code does seem fine. Just tested on OSX, havnt tested on windows though. It would be nice if you could open an issue on the repo to track it. Btw how are you generating the pem file ?

**Comment by kjlaw89** [comment]
@deepak1556 Okay, I opened up an issue for it. I'm not sure exactly what program we're using to generate the PEM files, but there's a pretty common Windows one. We normally create .p12 for our clients, which they just then install in the system certificate store and those work on all browsers. The .pem files are generated using the same method. Is there anything specific about the certificate that you want me to check or verify?

**Comment by deepak1556** [comment]
@kjlaw89 i was testing with https://github.com/thojansen/client-certificates/blob/master/ssl/setup.sh , except was importing .pem from .p12 using openssl after the last step (which is plain text with both Certificate and private key). Just to be sure is you PEM file encrypted ?

**Comment by kjlaw89** [comment]
I believe the process we use does the exact same thing, just behind the scenes. I do not believe the PEM file is encrypted, but I'd have to check to be certain. I do know that the PEM file I was using does not require a password, unlike the .p12 that I was using.

**Comment by kjlaw89** [comment]
Hey @deepak1556, have you had any luck fixing this? I'm starting full swing into my development now, so it'd be cool if this was resolved. I you need me to I can provide you a certificate that we generate with a limited access to our server so you can test with what we have.

Thanks again for your work on this! Don't forget to claim the bounty when it's all done!

**Comment by deepak1556** [comment]
@kjlaw89 didn't try since wasn't sure of what was missing in the certificate generation. Yeah it would be great if you could provide the certificate or the process of generation.

**Pull request merged**